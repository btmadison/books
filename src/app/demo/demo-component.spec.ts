import {
  DemoModule,
  BankAccountComponent, BankAccountParentComponent,
  LightswitchComponent,
  Child1Component, Child2Component, Child3Component,
  ExternalTemplateComponent,
  InputComponent,
  IoComponent, IoParentComponent,
  MyIfComponent, MyIfChildComponent, MyIfParentComponent,
  NeedsContentComponent, ParentComponent,
  TestProvidersComponent, TestViewProvidersComponent,
  ReversePipeComponent, ShellComponent
} from './demo';
import { Spectator, createComponentFactory, byText } from '@ngneat/spectator/jest';

// [1] unit testing components
// setup the component much less overhead with spectator (1.1)
// access the TS with spectator.component  (1.2)
// use spectator.detectChanges()  to trigger the change detection (1.3), (do not have to do it always)
// use DOM testing library convenience methods:  https://github.com/ngneat/spectator#queries' (1.4)


describe('Child1Component', () => {

  it('should create a component with inline template', () => {

    const spectator: Spectator<Child1Component> = createComponentFactory({
      component: Child1Component,
    })();

    expect(spectator.query(byText('Child'))).toBeDefined();
  });

  it('should create a component with external template', () => {

    const spectator: Spectator<ExternalTemplateComponent> = createComponentFactory({
      component: ExternalTemplateComponent,
    })();

    expect(spectator.query(byText('from external template'))).toBeDefined();
  });

  it('should allow changing members of the component', () => {

    const spectator: Spectator<MyIfComponent> = createComponentFactory({
      component: MyIfComponent,
      detectChanges: false
    })();
    const component = spectator.component;

    expect(spectator.query(byText('MyIf()'))).toBeDefined();

    component.showMore = true;
    spectator.detectChanges();
    expect(spectator.query(byText('MyIf(More)'))).toBeDefined();
  });


});
