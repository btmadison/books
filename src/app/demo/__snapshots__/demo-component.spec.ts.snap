// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[1] testing components, various examples ReversePipeComp should reverse the input text 1`] = `
<reverse-pipe-comp
  __ngContext__={[Function LRootView]}
  text={[Function String]}
>
  <input
    class="ng-untouched ng-valid ng-dirty"
  /><span>
    .xof nworb kciuq eht
  </span>
</reverse-pipe-comp>
`;

exports[`[1] testing components, various examples lifecycle hooks w/ MyIfParentComp changed parent value flows to child 1`] = `
<my-if-parent-comp
  __ngContext__={[Function LRootView]}
  ngOnInitCalled={[Function Boolean]}
  parentValue={[Function String]}
  showChild={[Function Boolean]}
  toggleLabel={[Function String]}
>
  <h3>
    MyIfParentComp
  </h3><label>
    Parent value: 
    <input
      class="ng-untouched ng-pristine ng-valid"
    />
  </label><button>
    Close Child
  </button><br /><div
    style="margin: 4px; padding: 4px; background-color: aliceblue;"
  >
    <my-if-child-1>
      <h4>
        MyIfChildComp
      </h4>
      <div>
        <label>
          Child value: 
          <input
            class="ng-untouched ng-pristine ng-valid"
          />
        </label>
      </div>
      <p>
        <i>
          Change log:
        </i>
      </p>
      <div>
        1 - value: currentValue = "Hello, World", previousValue = undefined
      </div>
      <div>
        2 - ngOnInit called
      </div>
      <div>
        3 - value: currentValue = "foo", previousValue = "Hello, World"
      </div>
      
    </my-if-child-1>
  </div>
</my-if-parent-comp>
`;

exports[`[1] testing components, various examples lifecycle hooks w/ MyIfParentComp child component should exist after OnInit, should have called child component OnInit 1`] = `
<my-if-parent-comp
  __ngContext__={[Function LRootView]}
  ngOnInitCalled={[Function Boolean]}
  parentValue={[Function String]}
  showChild={[Function Boolean]}
  toggleLabel={[Function String]}
>
  <h3>
    MyIfParentComp
  </h3><label>
    Parent value: 
    <input
      class="ng-untouched ng-pristine ng-valid"
    />
  </label><button>
    Close Child
  </button><br /><div
    style="margin: 4px; padding: 4px; background-color: aliceblue;"
  >
    <my-if-child-1>
      <h4>
        MyIfChildComp
      </h4>
      <div>
        <label>
          Child value: 
          <input
            class="ng-untouched ng-pristine ng-valid"
          />
        </label>
      </div>
      <p>
        <i>
          Change log:
        </i>
      </p>
      <div>
        1 - value: currentValue = "Hello, World", previousValue = undefined
      </div>
      <div>
        2 - ngOnInit called
      </div>
      
    </my-if-child-1>
  </div>
</my-if-parent-comp>
`;

exports[`[1] testing components, various examples lifecycle hooks w/ MyIfParentComp should instantiate parent component, OnInit should NOT be called before first detectChanges 1`] = `
<my-if-parent-comp
  __ngContext__={[Function LRootView]}
  ngOnInitCalled={[Function Boolean]}
  parentValue={[Function String]}
  showChild={[Function Boolean]}
  toggleLabel={[Function String]}
>
  <h3>
    MyIfParentComp
  </h3><label>
    Parent value: 
    <input
      class="ng-untouched ng-pristine ng-valid"
    />
  </label><button>
    Close Child
  </button><br /><div
    style="margin: 4px; padding: 4px; background-color: aliceblue;"
  >
    <my-if-child-1>
      <h4>
        MyIfChildComp
      </h4>
      <div>
        <label>
          Child value: 
          <input
            class="ng-untouched ng-pristine ng-valid"
          />
        </label>
      </div>
      <p>
        <i>
          Change log:
        </i>
      </p>
      <div>
        1 - value: currentValue = "Hello, World", previousValue = undefined
      </div>
      <div>
        2 - ngOnInit called
      </div>
      
    </my-if-child-1>
  </div>
</my-if-parent-comp>
`;

exports[`[1] testing components, various examples should allow changing members of the component 1`] = `
<my-if-comp
  __ngContext__={[Function LRootView]}
  showMore="false"
>
  MyIf()
</my-if-comp>
`;

exports[`[1] testing components, various examples should allow changing members of the component 2`] = `
<my-if-comp
  __ngContext__={[Function LRootView]}
  showMore={[Function Boolean]}
>
  MyIf(<span>
    More
  </span>)
</my-if-comp>
`;

exports[`[1] testing components, various examples should create a component with external template 1`] = `
<external-template-comp
  __ngContext__={[Function LRootView]}
  service="null"
>
  <span>
    from external template
  </span>
</external-template-comp>
`;

exports[`[1] testing components, various examples should create a component with inline template 1`] = `
<child-1
  __ngContext__={[Function LRootView]}
  text={[Function String]}
>
  <span>
    Child-1(Original)
  </span>
</child-1>
`;

exports[`[1] testing components, various examples should create a nested component bound to inputs/outputs 1`] = `
<io-parent-comp
  __ngContext__={[Function LRootView]}
  heroes={[Function Array]}
>
  <p>
    <i>
      Click to select a hero
    </i>
  </p><io-comp>
    <div
      class="hero"
    >
      Original Bob
    </div>
  </io-comp><io-comp>
    <div
      class="hero"
    >
      Original Carol
    </div>
  </io-comp><io-comp>
    <div
      class="hero"
    >
      Original Ted
    </div>
  </io-comp><io-comp>
    <div
      class="hero"
    >
      Original Alice
    </div>
  </io-comp>
</io-parent-comp>
`;

exports[`[1] testing components, various examples should create a nested component bound to inputs/outputs 2`] = `
<io-parent-comp
  __ngContext__={[Function LRootView]}
  heroes={[Function Array]}
  selectedHero={[Function Object]}
>
  <p>
    The selected hero is Bob
  </p><io-comp>
    <div
      class="hero"
    >
      Original Bob
    </div>
  </io-comp><io-comp>
    <div
      class="hero"
    >
      Original Carol
    </div>
  </io-comp><io-comp>
    <div
      class="hero"
    >
      Original Ted
    </div>
  </io-comp><io-comp>
    <div
      class="hero"
    >
      Original Alice
    </div>
  </io-comp>
</io-parent-comp>
`;

exports[`[1] testing components, various examples should support clicking a button 1`] = `
<lightswitch-comp
  __ngContext__={[Function LRootView]}
  isOn="false"
>
  <button>
    Click me!
  </button><span>
    The light is Off
  </span>
</lightswitch-comp>
`;

exports[`[1] testing components, various examples should support clicking a button 2`] = `
<lightswitch-comp
  __ngContext__={[Function LRootView]}
  isOn={[Function Boolean]}
>
  <button>
    Click me!
  </button><span>
    The light is On
  </span>
</lightswitch-comp>
`;

exports[`[1] testing components, various examples should support entering text in input box (ngModel) 1`] = `
<input-comp
  __ngContext__={[Function LRootView]}
  name={[Function String]}
>
  <input
    class="ng-untouched ng-pristine ng-valid"
  />
</input-comp>
`;

exports[`[1] testing components, various examples should support entering text in input box (ngModel) 2`] = `
<input-comp
  __ngContext__={[Function LRootView]}
  name={[Function String]}
>
  <input
    class="ng-untouched ng-pristine ng-valid"
  />
</input-comp>
`;
